services/ai_pr_review.py

```diff
@@ -0,0 +1,375 @@
+        return self._head_sha
+
+    async def fetch_review_comments(self):
+        async with self.torngit.get_client() as client:
+            page = 1
+            while True:
+                res = await self.torngit.api(
+                    client,
+                    "get",
+                    f"/repos/{self.torngit.slug}/pulls/{self.pullid}/comments?per_page=100&page={page}",
+                    token=self.token,
+                )
+                if len(res) == 0:
+                    break
+                for item in res:
+                    yield item
+                page += 1
+
+    async def create_review(self, commit_sha: str, review_comments: ReviewComments):
+        body = dict(
+            commit_id=commit_sha,
+            body=review_comments.body,
+            event="COMMENT",
+            comments=[
+                {
+                    "path": line_info.file_path,
+                    "position": line_info.position,
+                    "body": comment.body,
+                }
+                for line_info, comment in review_comments.comments.items()
+            ],
+        )
+        log.info(
+            "Creating AI PR review",
+            extra=body,
+        )
+
+        async with self.torngit.get_client() as client:
+            res = await self.torngit.api(
+                client,
+                "post",
+                f"/repos/{self.torngit.slug}/pulls/{self.pullid}/reviews",
+                token=self.token,
+                body=body,
+            )
+        return res
+
+    async def update_comment(self, comment: Comment):
+        log.info(
+            "Updating comment",
+            extra=dict(
+                comment_id=comment.comment_id,
+                body=comment.body,
+            ),
+        )
+        async with self.torngit.get_client() as client:
+            await self.torngit.api(
+                client,
+                "patch",
+                f"/repos/{self.torngit.slug}/pulls/comments/{comment.comment_id}",
+                token=self.token,
+                body=dict(body=comment.body),
+            )
+
+
+class Review:
+    def __init__(
+        self, pull_wrapper: PullWrapper, review_ids: Optional[List[int]] = None
+    ):
+        self.pull_wrapper = pull_wrapper
+        self.review_ids = review_ids or []
+        self.diff = None
+
+    async def perform(self) -> Optional[int]:
+        raw_diff = await self.pull_wrapper.fetch_diff()
+        self.diff = Diff(raw_diff)
+
+        prompt = build_prompt(self.diff.preprocessed)
+        log.debug(
+            "OpenAI prompt",
+            extra=dict(
+                prompt=prompt,
+            ),
+        )
+
+        res = await fetch_openai_completion(prompt)
+        log.debug(
+            "OpenAI response",
+            extra=dict(
+                res=res,
+            ),
+        )
+
+        try:
+            data = json.loads(res)
+        except json.decoder.JSONDecodeError:
```

Comment:
We should have metrics for these errors so we can track how successful this thing is (in terms of delivering any response, obviously it wouldn't measure how effective the response is)

Reply:
Added in https://github.com/codecov/worker/pull/187/commits/3de8687d6ec6de57b7bd0321b82dd21e29a5e20e